feat(assets): Add support for `srcset` and a Picture component (#8620)
withastro/astro: .changeset/smooth-goats-agree.md

---
'astro': minor
---

Add support for generating multiple widths when using the Image component and a Picture component for supporting multiple formats.

## `srcset` support

Two new properties have been added to `Image` and `getImage`: `densities` and `widths`.

> These props can be used to generate a `srcset` attribute with multiple sources. For example:

Author: The `width / 2` is actually required! Because densities should never upscale, so your base image needs to be smaller than the densities it generates. (You can think of `width` as controlling the base image, whereas `densities` control its variants)

If you have an image that is originally 1280x720, and you ask for density 2x (2560x1440) without specifying a smaller size for the base image, Astro would have to upscale the original image to reach that size, since we never upscale, `density={['2x']}` would just get ignored.

> These props can be used to generate a `srcset` attribute with multiple sources. For example:

Reviewer: I kind of know I knew that at one point... and I agree, no need to get into that here. But it doesn't seem intuitive to me that an array with a single value of 2 produces both a 2x and a 3x in `srcset`. Admittedly, if I understood this better, I'd be better able to say what should/shouldn't go in a description here. Naively, I would have expected to need to specify two values in my array to get two generated.

> These props can be used to generate a `srcset` attribute with multiple sources. For example:

Author: Oh, it doesn't sorry!

If you do `densities={[2]}`, it'll only have the original density (in the `src` attribute) and the 2x in `srcset`.

```astro
---
import { Image } from "astro";
import myImage from "./my-image.jpg";
---

<Image src={myImage} width={myImage.width / 2} densities={[2]} alt="My cool image"  />
```

```html
<img src="..." srcset="... 2x, ... 3x" alt="My cool image" />
```

## Picture component

> The `Picture` component can be used to generate a `<picture>` element with multiple sources. It can be used as follow:

Reviewer: ```suggestion
The experimental `<Picture />` component can be used to generate a `<picture>` element with multiple `<source>` elements. 

The example below uses `format` property to generate a `<source>` in each of the specified image formats:
```
I kind of like making "sources" code here, to distinguish the tag from "multiple sources" sounding like "original sources/locations/files" etc. unless there's a good reason not to do this! (And if "elements" isn't the best word, pick another word like tag or attribute or whatever is technically accurate here. It's also ok just to pluralize as `<source>`s, but that doesn't always visually look the best.)

Also, should probably be re-upping experimental, right?  :sweat_smile:

```astro
---
import { Picture } from "astro:assets";
import myImage from "./my-image.jpg";
---

<Picture src={myImage} formats={["avif", "webp"]} alt="My super image in multiple formats!" />
```

> The above code will generate the following:

Reviewer: ```suggestion
The above code will generate the following HTML, and allow the browser to determine the best image to display:
```
I originally came just to add "HTML", but since this will be also material for the blog post draft, I think throwing in a benefit of some kind at the end of the sentence will be helpful for the release blog post AND gives us a bit of a nicer sentence here.

Pick whatever benefit/description you want, but I like a beefier sentence showing off how this is useful here!

---
'astro': minor
---

Adds experimental support for generating `srcset` attributes and a new `<Picture />` component.

## `srcset` support

Two new properties have been added to `Image` and `getImage()`: `densities` and `widths`.

> These properties can be used to generate a `srcset` attribute, either based on absolute widths in pixels (e.g. [300, 600, 900]) or pixel density descriptors (e.g. `["2x"]` or `[1.5, 2]`).

Reviewer: love the pixel density descriptors showing both unit options!

```astro
---
import { Image } from "astro";
import myImage from "./my-image.jpg";
---

<Image src={myImage} width={myImage.width / 2} densities={[1.5, 2]} alt="My cool image"  />
```

```html
<img
  src="/_astro/my_image.hash.webp"

>   srcset="/_astro/my_image.hash.webp 1.5x, /_astro/my_image.hash.webp 2x"

Reviewer: These look way more instructive now! :raised_hands:

  alt="My cool image"
/>
```

## Picture component

The experimental `<Picture />` component can be used to generate a `<picture>` element with multiple `<source>` elements.

> The example below uses `format` property to generate a `<source>` in each of the specified image formats:

Reviewer: ```suggestion
The example below uses the `format` property to generate a `<source>` in each of the specified image formats:
```
grammar nit
